bittensor/_dataset/dataset_impl.py

```diff
@@ -1,536 +1,490 @@
-            status_forcelist=(104, 500, 502, 504),
-            session=None,
-        ):
-        """ Creates a retriable session for request calls. This enables
-        automatic retries and back-off retries should any request calls fail.
 
+    def __init__(self,  max_threads:int=None):
+        """Initializes a new ThreadPoolExecutor instance.
         Args:
-            retries (int, optional): Maximum number of retries. Defaults to 3.
-            backoff_factor (float, optional): Factor by which to back off if a retry fails. Defaults to 0.3.
-            status_forcelist (tuple, optional): A set of integer HTTP status codes that we should force a retry on. Defaults to (500, 502, 504).
-            session ([type], optional): Session for which to set up the retries. Defaults to None.
-
-        Returns:
-            requests.Session(): A Requests Session object set up for retries and backoff.
-        """
-        session = session or requests.Session()
-        retry = Retry(
-            total=retries,
-            read=retries,
-            connect=retries,
-            backoff_factor=backoff_factor,
-            status_forcelist=status_forcelist,
-        )
-        adapter = HTTPAdapter(max_retries=retry)
-        session.mount('http://', adapter)
-        session.mount('https://', adapter)
-        return session
-
-    def get_ipfs_directory(self, address: str, file_meta: dict, action: str = 'post', timeout : int = 180):
-        r"""Connects to IPFS gateway and retrieves directory.
-        Args:
-            address: (:type:`str`, required):
-                The target address of the request. 
-            params: (:type:`tuple`, optional):
-                The arguments of the request. eg. (('arg', dataset_hash),)
-            action: (:type:`str`, optional):
-                POST or GET.
-            timeout: (:type:`int`, optional):
-                Timeout for getting the server's response. 
-        Returns:
-            dict: A dictionary of the files inside of the genesis_datasets and their hashes.
+            max_threads: 
+                The maximum number of threads that can be used to
+                execute the given calls.
         """
-        session = requests.Session()
-        session.params.update((('arg', file_meta['Hash']), ))
-        
-        try:
-            if action == 'get':
-                response = self.requests_retry_session(session=session).get(address, timeout=timeout)
-            elif action == 'post':
-                response = self.requests_retry_session(session=session).post(address, timeout=timeout)
-            logger.success("Loaded from IPFS:".ljust(20) + "<blue>{}</blue>".format(file_meta['Name']))
+        self.max_threads = max_threads
+        self._idle_semaphore = threading.Semaphore(0)
+        self._threads = []
+        self._shutdown_lock = threading.Lock()
+        self._shutdown = False
+
+    def submit(self, fn, args:list=[],kwargs:dict={}):
```

Comment:
Default args shouldn't be mutable. Should be `Optional[list] = None` and instantiate it as `[]` within the function. (same for the kwargs)

Reply:
Or just use the standard `*args` and `**kwargs`, should behave fine since they become a tuple and a dict within here.